Framework/Source/Graphics/Program/ParameterBlock.cpp

```diff
@@ -0,0 +1,776 @@
+/***************************************************************************
+# Copyright (c) 2015, NVIDIA CORPORATION. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions
+# are met:
+#  * Redistributions of source code must retain the above copyright
+#    notice, this list of conditions and the following disclaimer.
+#  * Redistributions in binary form must reproduce the above copyright
+#    notice, this list of conditions and the following disclaimer in the
+#    documentation and/or other materials provided with the distribution.
+#  * Neither the name of NVIDIA CORPORATION nor the names of its
+#    contributors may be used to endorse or promote products derived
+#    from this software without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS ``AS IS'' AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+# PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNER OR
+# CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+# EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+# PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+# OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+***************************************************************************/
+#include "Framework.h"
+#include "ParameterBlock.h"
+#include "API/LowLevel/RootSignature.h"
+#include "API/Device.h"
+
+namespace Falcor
+{
+    void bindConstantBuffers(const ParameterBlock::ResourceMap<ConstantBuffer>& cbMap, const ParameterBlock::RootSetVec& rootSets);
+    ReflectionResourceType::ShaderAccess getRequiredShaderAccess(RootSignature::DescType type);
+
+    ParameterBlock::RootData findRootData(const RootSignature* pRootSig, uint32_t regIndex, uint32_t regSpace, RootSignature::DescType descType)
+    {
+        // Search the descriptor-tables
+        for (size_t i = 0; i < pRootSig->getDescriptorSetCount(); i++)
+        {
+            const RootSignature::DescriptorSetLayout& set = pRootSig->getDescriptorSet(i);
+            for (uint32_t r = 0; r < set.getRangeCount(); r++)
+            {
+                const RootSignature::DescriptorSetLayout::Range& range = set.getRange(r);
+                if (range.type == descType && range.regSpace == regSpace)
+                {
+                    if (range.baseRegIndex == regIndex)
```

Comment:
Should this not just be testing `baseRegIndex`, but instead whether the `regIndex` is between `range.baseRegIndex` and `range.count` (exclusive)?